File: examples/haskell/BoundedSlice.hs
-- A slice that REFUSES an out-of-range request instead of returning a shorter one.
--
-- The defect this kills: `take n . drop k` silently returns fewer elements than asked for when
-- the list is too short. The caller gets a list, and a list of the wrong length is
-- indistinguishable from a list of the right one unless somebody checks — so the truncation
-- travels as a result.
--
-- Verify: runghc examples/haskell/BoundedSlice.hs
module Main (main) where

import Control.Monad (unless)
import System.Exit (exitFailure)

-- Nothing is the REFUSAL, and it is a different constructor from a short list, so the two cannot
-- be confused by a caller that forgot to check a length.
slice :: Int -> Int -> [a] -> Maybe [a]
slice offset count xs
  | offset < 0 || count < 0            = Nothing
  | offset + count > length xs         = Nothing   -- never a clamp
  | otherwise                          = Just (take count (drop offset xs))

check :: String -> Bool -> IO ()
check name ok = unless ok $ putStrLn ("FAILED: " ++ name) >> exitFailure

main :: IO ()
main = do
  let xs = [0 .. 7] :: [Int]
  check "whole"          (slice 0 8 xs == Just xs)
  check "middle"         (slice 2 3 xs == Just [2, 3, 4])
  check "count past end" (slice 0 9 xs == Nothing)
  check "offset past end"(slice 9 0 xs == Nothing)
  check "valid offset, invalid count" (slice 6 4 xs == Nothing)
  -- An empty result is a real answer and is NOT a refusal; the two must stay distinguishable.
  check "empty is not refusal" (slice 8 0 xs == Just [])
  check "negative refused"  (slice (-1) 1 xs == Nothing)
  putStrLn "BoundedSlice: 7 assertions held - refusal, empty, and a negative that does not wrap"

File: examples/ocaml/bounded_slice.ml
(* A slice that REFUSES out-of-range bounds instead of raising deep inside a caller.

   The defect this kills: String.sub raising Invalid_argument three frames away from the bad
   index, where nobody knows which input produced it. Returning an option makes the refusal a
   value the caller must handle, at the call site.

   Verify: ocaml examples/ocaml/bounded_slice.ml *)

let slice s ~pos ~len =
  if pos < 0 || len < 0 || pos > String.length s - len then None
  else Some (String.sub s pos len)

let () =
  assert (slice "heartland" ~pos:0 ~len:5 = Some "heart");
  assert (slice "heartland" ~pos:5 ~len:4 = Some "land");
  assert (slice "heartland" ~pos:6 ~len:4 = None);   (* one past the end: refused, not raised *)
  assert (slice "heartland" ~pos:(-1) ~len:2 = None);
  assert (slice "" ~pos:0 ~len:0 = Some "");
  print_endline "bounded_slice: ok"
